ts/src/profiler.ts

```diff
@@ -212,24 +300,27 @@ export class Profiler extends common.ServiceObject {
    *
    * @return time, in ms, to wait before asking profiler server again about
    * collecting another profile.
-   *
-   * TODO: implement backoff and retry. When error encountered in
-   * createProfile() should be retried when response indicates this request
-   * should be retried or with exponential backoff (up to one hour) if the
-   * response does not indicate when to retry this request.
    */
   async collectProfile(): Promise<number> {
     let prof: RequestProfile;
     try {
       prof = await this.createProfile();
     } catch (err) {
-      this.logger.error(`Failed to create profile: ${err}`);
-      return this.config.backoffMillis;
+      if (isBackoffResponseError(err)) {
+        this.logger.debug(`Must wait ${
+            msToStr(err.backoffMillis)} to create profile: ${err}`);
+        return Math.min(err.backoffMillis, this.config.serverBackoffCapMillis);
+      }
+      const backoff = this.retryer.getBackoff();
+      this.logger.warn(`Failed to create profile, waiting ${
+          msToStr(backoff)} to try again: ${err}`);
+      return backoff;
     }
+    this.retryer.reset();
     try {
       await this.profileAndUpload(prof);
     } catch (err) {
-      this.logger.error(`Failed to collect and upload profile: ${err}`);
+      this.logger.warn(`Failed to collect and upload profile: ${err}`);
```

Comment:
nit: s/collect and upload/collect or upload/? It would be nice if we could specify what failed precisely. Perhaps that can be fixed by adding logging statements inside `profileAndUpload`.

Reply:
I've shifted the logging to profileAndUpload. 
Hopefully code is still clear.